Source/WebKit/NetworkProcess/storage/NetworkStorageManager.cpp

```diff
@@ -624,8 +664,15 @@ void NetworkStorageManager::persist(const WebCore::ClientOrigin& origin, Complet
     if (persistedFilePath(origin).isEmpty())
         return completionHandler(false);
 
-    // FIXME: add heuristics to decide if origin can be persisted and write persited file if it can.
-    return completionHandler(false);
+    bool shouldFetchRegistrableDomains = m_persistCompletionHandlers.isEmpty();
+    m_persistCompletionHandlers.append({ origin, WTFMove(completionHandler) });
+    if (!shouldFetchRegistrableDomains)
+        return;
+
+    RunLoop::main().dispatch([weakThis = ThreadSafeWeakPtr { *this }]() mutable {
+        if (auto strongThis = weakThis.get())
+            strongThis->fetchRegistrableDomainsForPersist();
```

Comment:
The issue is that the registrable domains might be different between the time persist is called by the JS app (at that point, it could be exempt from ITP deletion) and when the storage pressure happens (say a few days later) at which point this domain might no longer be exempt from ITP deletion.
In that case, what should happen? Should we delete the data, defeating persistency, or keep the data, defeating ITP?

Also, this API would expose whether a website is exempt from ITP deletion or not. So far, I do not think we expose that bit of information to web pages, that could be used as a fingerprint information.

Another approach could be:
- Add a callback to application to ask whether a domain should be persisted or not. Default implementation might be yes. If doing so, we might need to surface to the application the persisting boolean flag.
- When storage pressure happens, clear everything possible except for storage marked as persisting.
- If this is not enough, expose a callback to notify the application that storage is too big and that it should take some actions (including clearing persisting data). I am not sure what the default implementation should be if the callback is not implemented, maybe delete persisting data by last time domain was visited?

Reply:
Discussed on slack. (Domains exempt from ITP deletion are configured by WebKit clients, which should be fixed; they are not equal to domains that ITP decides to keep data, which can change during runtime).